File: game.py
import random
import os
import time

size = os.get_terminal_size()

field = [[0 for i in range(size[0])] for i in range(size[1])]
output = [[0 for i in range(size[0])] for i in range(size[1])]
duplicate = [[0 for i in range(size[0])] for i in range(size[1])]

# patterns
patterns = {
    0: [[1, 1],
        [1, 1]], # block
    1: [[0, 1, 1, 0],
        [1, 0, 0, 1],
        [0, 1, 1, 0]], #beehive
    2: [[0, 1, 1, 0],
        [1, 0, 0, 1],
        [0, 1, 0, 1],
        [0, 0, 1, 0]], # loaf
    3: [[1, 1, 0],
        [1, 0, 1],
        [0, 1, 0]], # boat
    4: [[0, 1, 0],
        [1, 0, 1],
        [0, 1, 0]], # tub
    5: [[1],
        [1],
        [1]], # blinker
    6: [[0, 1, 1, 1],
        [1, 1, 1, 0]], # toad
    7: [[1, 1, 0, 0],
        [1, 1, 0, 0],
        [0, 0, 1, 1],
        [0, 0, 1, 1]], # beacon
    8: [[0, 1, 0],
        [0, 0, 1],
        [1, 1, 1]], # glider
    9: [[0, 1, 1],
        [1, 1, 0],
        [0, 1, 0]], # r-pentomino
    10: [[0, 0, 0, 0, 0, 0, 1, 0],
         [1, 1, 0, 0, 0, 0, 0, 0],
         [0, 1, 0, 0, 0, 1, 1, 1]], # diehard
    11: [[0, 1, 0, 0, 0, 0, 0],
         [0, 0, 0, 1, 0, 0, 0],
         [1, 1, 0, 0, 1, 1, 1]], # acorn
    12: [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
         [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]] # gosper glider gun

}

def initialize(field):
    number = random.randint(4, 12)

    for i in range(number):
        n = random.randint(0, len(patterns)-1)

        if i >= number//2:
            n = random.randint(7, len(patterns)-1)

        x = random.randint(0, int(size[0]))
        y = random.randint(0, int(size[1]))

        height = len(patterns[n])
        width = len(patterns[n][0])

        for j in range(height):
            for k in range(width):

                ypos = y + j
                xpos = x + k

                if ypos > size[1] - 1:
                    ypos = j
                if xpos > size[0] - 1:
                    xpos = k

                if j > size[1] - 1:
                    ypos = 0
                if k > size[0] - 1:
                    xpos = 0

                field[ypos][xpos] = patterns[n][j][k]
                duplicate[ypos][xpos] = patterns[n][j][k]


initialize(field)

def count(field, y, x):
    neighbors = 0
    
    x1 = x + 1
    y1 = y + 1

    if x1 >= len(field[y]):
        x1 = 0
    if y1 >= len(field):
        y1 = 0
        
    cells = [field[y-1][x-1], field[y-1][x], field[y-1][x1], field[y][x-1], field[y][x1], field[y1][x-1], field[y1][x], field[y1][x1]]

    neighbors = cells.count(1)

    return neighbors

    

while True:
    output = [["#" if x == 1 else " " for x in row] for row in field]


    outlist = [''.join(str(c) for c in lst) for lst in output]
    outl = '\n'.join(outlist)

    print(outl)
    time.sleep(0.22)

    for y in range(len(field)):
        for x in range(len(field[y])):
            neighbors = count(duplicate, y, x)
            
            if neighbors == 3:
                field[y][x] = 1
            elif neighbors < 2 or neighbors > 3:
                field[y][x] = 0
            

    for y in range(len(field)):
        for x in range(len(field[y])):
            duplicate[y][x] = field[y][x]

File: game.rb
# Conway's Game of Life

# specific how many octaves, lowest frequency, scale, synths, volume, note length
num = 1
lf = 60
s = :minor_pentatonic
livesynth = :tri
deadsynth = :fm
livevol = 1
deadvol = 0.5
d = 0.3

# patterns
patterns = {
  
  0 => [[1, 1],
        
        [1, 1]], # block
  
  1 => [[0, 1, 1, 0],
        
        [1, 0, 0, 1],
        
        [0, 1, 1, 0]], #beehive
  
  2 => [[0, 1, 1, 0],
        
        [1, 0, 0, 1],
        
        [0, 1, 0, 1],
        
        [0, 0, 1, 0]], # loaf
  
  3 => [[1, 1, 0],
        
        [1, 0, 1],
        
        [0, 1, 0]], # boat
  
  4 => [[0, 1, 0],
        
        [1, 0, 1],
        
        [0, 1, 0]], # tub
  
  5 => [[1],
        
        [1],
        
        [1]], # blinker
  
  6 => [[0, 1, 1, 1],
        
        [1, 1, 1, 0]], # toad
  
  7 => [[1, 1, 0, 0],
        
        [1, 1, 0, 0],
        
        [0, 0, 1, 1],
        
        [0, 0, 1, 1]], # beacon
  
  8 => [[0, 1, 0],
        
        [0, 0, 1],
        
        [1, 1, 1]], # glider
  
  9 => [[0, 1, 1],
        
        [1, 1, 0],
        
        [0, 1, 0]], # r-pentomino
  
  10 => [[0, 0, 0, 0, 0, 0, 1, 0],
         
         [1, 1, 0, 0, 0, 0, 0, 0],
         
         [0, 1, 0, 0, 0, 1, 1, 1]], # diehard
  
  11 => [[0, 1, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 1, 0, 0, 0],
         
         [1, 1, 0, 0, 1, 1, 1]], # acorn
  
  12 => [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
         
         [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
         
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]] # gosper glider gun
}

# plays the game
define :keysmith do |st, n|
  sc = (scale lf, s)
  sl = (scale lf, s, num_octaves: n)
  k = []
  i = 0
  (sc.length() - 1).times do
    k << sc[i]
    i += 1
  end
  
  m = []
  i = 0
  (sl.length() - 1).times do
    m << sl[i]
    i += 1
  end
  
  l = m.length()
  
  return k.ring, l, m.ring
end

keys, l, scales = keysmith lf, num

define :init do
  field = []
  duplicate = []
  l.times do
    row = []
    drow = []
    l.times do
      row << 0 #rrand_i(0,1)
      drow << 0
    end
    field << row
    duplicate << drow
  end
  
  return field, duplicate
end

y0, dy = init

define :place do |p|
  i = 0
  p.length().times do
    j = 0
    p[0].length().times do
      y0[i][j] = p[i][j]
      dy[i][j] = p[i][j]
      j += 1
    end
    i+=1
  end
end

place patterns[8]

define :foo do |key, i, j, life|
  ai = ((100.0 - 100*i/l)/100.0) * life
  aj = ((100.0 - 100*j/l)/100.0) * life
  
  play scales[i] + (key - lf), amp: ai
  play scales[j] + (key - lf), amp: aj
  sleep d
end

define :count do |x, y|
  if x == l
    x = 0
  end
  
  if y == l
    y = 0
  end
  
  i = y0[y][x]
  
  return i
end

define :neighbor do |x, y|
  n = count x-1, y-1
  n += count x, y-1
  n += count x+1, y-1
  n += count x-1, y
  n += count x+1, y
  n += count x-1, y+1
  n += count x, y+1
  n += count x+1, y+1
  
  return n
end

define :game do
  i = 0
  l.times do
    j = 0
    l.times do
      n = neighbor j, i
      if n == 3
        dy[i][j] = 1
      end
      
      if n > 3
        dy[i][j] = 0
      end
      
      if n < 2
        dy[i][j] = 0
      end
      
      j += 1
    end
    
    i += 1
  end
  
  i = 0
  l.times do
    j = 0
    l.times do
      y0[i][j] = dy[i][j]
      j += 1
    end
    i+=1
  end
end

loop do
  key = keys.tick
  i = 0
  l.times do
    j = 0
    l.times do
      if y0[i][j] == 1
        use_synth livesynth
        foo key, i, j, livevol
      else
        use_synth deadsynth
        foo key, i, j, deadvol
      end
      j += 1
    end
    i += 1
  end
  game
end